File: solutions/elixir/proverb/1/lib/proverb.ex
defmodule Proverb do
  @doc """
  Generate a proverb from a list of strings.
  """

  @spec recite(strings :: [String.t()]) :: String.t()
  def recite([]), do: ""
  def recite([string]), do: "And all for the want of a #{string}.\n"

  def recite([first | rest] = strings) do
    lines =
      Enum.zip(strings, rest)
      |> Enum.map_join("\n", fn {prev, cur} -> "For want of a #{prev} the #{cur} was lost." end)

    String.trim_leading("""
    #{lines}
    And all for the want of a #{first}.
    """)
  end
end

File: solutions/typescript/proverb/1/proverb.ts
const chunk = (array: string[], chunkSize: number) => {
  var retArr: string[][] = [];
  for (let i = 0; i < array.length - (chunkSize - 1); i++) {
      retArr.push(array.slice(i, i + chunkSize));
  }
  return retArr;
}

export function proverb(...input: string[]): string {
  const chunks = chunk(input, 2);
  const ret: string[] = chunks.map(([a, b]: string[]): string => (
    `For want of a ${a} the ${b} was lost.`
  ));

  return ret.join("\n") + "\n" + `And all for the want of a ${input[0]}.`;
}
